aten/src/ATen/core/op_registration/op_registration.h

```diff
@@ -240,6 +255,34 @@ class CAFFE2_API RegisterOperators final {
       );
     }
 
+    /**
+     * Use this to register an operator whose kernel is implemented by a TORCH_FN.
+     * The kernel is only called for inputs matching the given dispatch key.
+     * You can register multiple kernels for different dispatch keys.
+     *
+     * Example:
+     *
+     * > namespace { Tensor my_kernel_cpu(Tensor a, Tensor b) {...} }
+     * >
+     * > static auto registry = c10::RegisterOperators()
+     * >     .op(c10::RegisterOperators::options()
+     * >         .schema("my_op")
+     * >         .kernel(DispatchKey::CPU, TORCH_FN(my_kernel_cpu)));
+     */
+    template<class FuncPtr>
+    // enable_if: only enable it if FuncType is actually a function
+    std::enable_if_t<c10::is_compile_time_function_pointer<FuncPtr>::value, Options&&> kernel(DispatchKey dispatch_key, FuncPtr) && {
```

Comment:
Why are you adding more API surface to RegisterOperators? We should add `TORCH_FN` support to `torch/library.h` only and just port any sites which we want to use the new functionality.

Reply:
I didn't want to port all call sites to the new API, that would have been an even larger PR. I need TORCH_FN for those cases though, so I added it to the old API too. Half of the functions I added here are scheduled to be removed very soon though (the unboxedOnly ones), so it's not that much of an API surface increase.

Reply:
Hmm, I'd echo not wanting to extend the old API. Unless TORCH_FN is urgent for some reason (is it? I actually thought it was an H2 thing) I think a natural dependency ordering is to port everything to the new API before doing anything that would require changes to the old.

I agree that the porting belongs in a different PR though. I mean, we should really do that anyway, regardless of new functionality. How much work is porting them, should we divvy them up to get them done?

Reply:
I need TORCH_FN for the unboxing metaprogramming, that's why I did it now instead of waiting until we ported the API.